include/ur_api.h

```diff
@@ -987,6 +989,70 @@ typedef enum ur_adapter_backend_t {
 
 } ur_adapter_backend_t;
 
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Minimum level of messages to be processed by the logger.
+typedef enum ur_logger_level_t {
+    UR_LOGGER_LEVEL_DEBUG = 0, ///< Debugging messages used for development purposes.
+    UR_LOGGER_LEVEL_INFO = 1,  ///< General messages not related to debugging, warnings or errors.
+    UR_LOGGER_LEVEL_WARN = 2,  ///< Used to warn users about potential problems.
+    UR_LOGGER_LEVEL_ERR = 3,   ///< Used when an error has occurred.
+    UR_LOGGER_LEVEL_QUIET = 4, ///< Restrict logger processing any messages.
+    /// @cond
+    UR_LOGGER_LEVEL_FORCE_UINT32 = 0x7fffffff
+    /// @endcond
+
+} ur_logger_level_t;
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Callback function to retrieve output from the logger.
+typedef void (*ur_logger_callback_t)(
+    ur_logger_level_t level, ///< [out] Minimum level of messages to be processed by the logger.
+    const char *pLoggerMsg,  ///< [in][out] pointer to data to be passed to callback
+    void *pUserData          ///< [in][out] pointer to data to be passed to callback
+);
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Set a callback function for use by the logger to retrieve logging output.
+///        It is a requirement that the callback function is thread safe and the
+///        creator of the function will be responsible for this.
+///
```

Comment:
One more thing to document is that calling this, `urAdapterSetLoggerCallback`, and `urAdapterSetLoggerCallbackLevel` functions is not thread-safe with respect to any other function in the API. The reason is that they are modifying global state without using atomics or grabbing a lock. In other words, when something is writing a log, you may invertedly change the logger callback pointer.

Reply:
Okay no problem, I reworded the description for `urAdapterSetLoggerCallback` but will add it to `urAdapterSetLoggerCallbackLevel` as well. It should specifically say not "thread safe with respect to any other function in the API"? Should the explanation be added here or is that enough?

Reply:
Most people understand thread-safe functions as safe to be invoked concurrently by multiple threads. And that is indeed how POSIX defines it. In this case, the situation is a bit different, since `urAdapterSetLoggerCallback` is not safe to call even with respect to other UR API functions (that are otherwise thread-safe). Small caveat is that implementations may choose to limit this unsafety to be per-adapter.

In my opinion this break [the principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), and should be thoroughly documented. Saying "not thread-safe" is not enough.

Ideally we'd fix this - one approach would be to only allow setting the callback as part of the loader config, and then change the logger level variable to be an atomic.

Reply:
Thanks Piotr. @kbenzie do you have any thoughts on this? Would we want to modify the design now or just get this PR in as it is now and modify it later? If the latter then I'll move these comments to a new issue to track this.